File: solutions/cpp/Picking_up_coins_dont_lose.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <numeric>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
// Return 0 means choosing F (even numbered coins),
// and return 1 means choosing S (odd numbered coins).
int pick_up_coins(const vector<int>& C) {
  int even_sum = 0, odd_sum = 0;
  for (int i = 0; i < C.size(); ++i) {
    if (i % 2) {  // odd.
      odd_sum += C[i];
    } else {  // even.
      even_sum += C[i];
    }
  }
  return even_sum >= odd_sum ? 0 : 1;
}
// @exclude

void check(const vector<int>& C, int choose) {
  int even = 0, odd = 0;
  for (int i = 0; i < C.size(); ++i) {
    if (i % 2) {
      odd += C[i];
    } else {
      even += C[i];
    }
  }
  if (choose == 0) {
    assert(even >= odd);
  } else {
    assert(odd >= even);
  }
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    vector<int> C;
    if (argc >= 2) {
      for (int i = 1; i < argc; ++i) {
        C.emplace_back(atoi(argv[i]));
      }
    } else {
      uniform_int_distribution<int> dis(1, 1000);
      C.resize(dis(gen));
      for (int i = 0; i < C.size(); ++i) {
        uniform_int_distribution<int> dis(0, 99);
        C[i] = dis(gen);
      }
    }

    for (size_t i = 0; i < C.size(); ++i) {
      cout << C[i] << ' ';
    }
    cout << endl;

    int res = pick_up_coins(C);
    cout << res << endl;
    check(C, res);
  }
  return 0;
}

File: solutions/java/src/main/java/com/epi/PickingUpCoinsDontLose.java
package com.epi;

import java.util.Random;

import static com.epi.utils.Utils.simplePrint;

public class PickingUpCoinsDontLose {
  // @include
  // Return 0 means choosing F (even numbered coins),
  // and return 1 means choosing S (odd numbered coins).
  static int pickUpCoins(int[] C) {
    int evenSum = 0, oddSum = 0;
    for (int i = 0; i < C.length; ++i) {
      if ((i % 2) == 1) { // odd.
        oddSum += C[i];
      } else { // even.
        evenSum += C[i];
      }
    }
    return evenSum >= oddSum ? 0 : 1;
  }
  // @exclude

  static void check(int[] C, int choose) {
    int even = 0, odd = 0;
    for (int i = 0; i < C.length; ++i) {
      if ((i % 2) == 1) {
        odd += C[i];
      } else {
        even += C[i];
      }
    }
    if (choose == 0) {
      assert(even >= odd);
    } else {
      assert(odd >= even);
    }
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int[] C;
      if (args.length >= 1) {
        C = new int[args.length - 1];
        for (int i = 0; i < args.length; ++i) {
          C[i] = Integer.valueOf(args[i]);
        }
      } else {
        C = new int[gen.nextInt(1000)];
        for (int i = 0; i < C.length; ++i) {
          C[i] = gen.nextInt(100);
        }
      }

      simplePrint(C);
      System.out.println();

      int res = pickUpCoins(C);
      System.out.println(res);
      check(C, res);
    }
  }
}
